### PR TITLE
Make builds more Node-friendly, convert project to ESM

### PR DESCRIPTION
**Pull request by kherock**
Using this project in Node is a bit strange in how it adds `fetch` to the global namespace and additionally installs a few dependencies that aren't actually needed for the project to be used as a module. Since this project is already using Babel, it's also not too much effort to update modules to use ESM. None of these changes should cause any breaking changes to the public API.

These changes enable tree shaking for webpack builds and avoids polluting the global namespace for Node.js without impacting the 'pure' builds. I've also added the `Client` class the public module interface so that it can be instantiated outside of Express middleware.

### CODE REVIEW

**Comment by vlad-ignatov** [comment]
Thank you so much @kherock ! This is a big change and I want to test it carefully. For now it seems to work and I only got a couple of TypeScript validation issues that I will describe separately as code comments.

File: src/Client.js [hunk 1]
```diff
@@ -12,15 +12,15 @@ const {
...
-const debug = _debug.extend("client");
-const str = require("./strings");
-const { fetchConformanceStatement, fetchFhirVersion } = require("./smart");
-const { SMART_KEY, patientCompartment, fhirVersions } = require("./settings");
+import str from "./strings";
+import { fetchConformanceStatement, fetchFhirVersion } from "./smart";
+import { SMART_KEY, patientCompartment, fhirVersions } from "./settings";
 
 // eslint-disable-next-line no-undef
 const { Response } = global.FHIRCLIENT_PURE ? window.fetch : require("cross-fetch");
```
Comment (vlad-ignatov) [inline src/Client.js]: It seems to me that this should be:
```js
const { Response } = global.FHIRCLIENT_PURE ? window : require("cross-fetch");
```

File: src/smart.js [hunk 2]
```diff
@@ -1,30 +1,21 @@
...
- * Note that this is done within a function to postpone the "./Client" import
- * and avoid cyclic dependency.
- * @param {fhirclient.JsonObject} env The adapter
- * @param {string | fhirclient.ClientState} state The client state or baseUrl
- * @returns {fhirclient.Client}
- */
-function createClient(env, state) {
-    const Client  = require("./Client");
-    return new Client(env, state);
-}
```
Comment (vlad-ignatov) [inline src/smart.js]: I vaguely remember that the reason for the existence of the (otherwise useless) `createClient` function was to avoid some kind of circular dependency, e.g.: `Client.js` and `smart.js` require each other. Perhaps this is not an issue for ESM but I have to ask - have you had any problems with that?

File: dist/lib/Client.js [hunk 3]
```diff
@@ -45,7 +45,12 @@ const {
   SMART_KEY,
   patientCompartment,
   fhirVersions
-} = require("./settings");
+} = require("./settings"); // eslint-disable-next-line no-undef
+
+
+const {
```
Comment (vlad-ignatov) [inline dist/lib/Client.js]: I have some issues with this statement:
1. TypeScript is unable to understand `global.FHIRCLIENT_PURE`. The same applies to the other files where it is used. So far my fix is to add this in `types.d.ts`:
    ```ts
    declare module NodeJS {
        interface Global {
            FHIRCLIENT_PURE?: Boolean 
        }
    }
    ```
    I'm open to suggestions if you know another way to fix this.
2. In browsers, `Response` is a global object so the second part should be `window` instead of `window.fetch`. In other words:
```js
const { Response } = global.FHIRCLIENT_PURE ? window : require("cross-fetch");
```

File: dist/lib/lib.js [hunk 4]
```diff
@@ -16,7 +16,10 @@ const debug = require("debug")("FHIR");
 
 const {
   patientParams
-} = require("./settings");
+} = require("./settings"); // eslint-disable-next-line no-undef
+
+
+const fetch = global.FHIRCLIENT_PURE ? window.fetch : require("cross-fetch");
```
Comment (vlad-ignatov) [inline dist/lib/lib.js]: I think this should be
```js
const fetch = global.FHIRCLIENT_PURE ? window.fetch : require("cross-fetch").fetch;
```
Even though `const fetch = require("cross-fetch");` is technically valid, TS won't understand it unless we use `require("cross-fetch").fetch` or `require("cross-fetch").default`.

Also, what is the reason for switching to `cross-fetch`? I see your point for not polluting the global namespace, but in old browsers without fetch support that is exactly what we should do. In other words, this library should act as polyfill in this case and provide `Request`, `Response`, `Headers`, `fetch`, `URL` and `URLSearchParams` to the global scope. Is that the case now?

File: src/Client.js (hunk 1 above)
Comment (kherock) [reply src/Client.js]: Oops, yes you are correct!

File: src/smart.js (hunk 2 above)
Comment (kherock) [reply src/smart.js]: ESM can resolve the circular import by the time the other functions in this file need it. It was necessary as CJS code prior to the conversion.

File: dist/lib/Client.js (hunk 3 above)
Comment (kherock) [reply dist/lib/Client.js]: 1. I think it depends on if we want this variable to be available outside the library? It's only a webpack build detail, so if this were converted to TypeScript code, I would probably write it `(global as any).FHIRCLIENT_PURE`.

File: dist/lib/lib.js (hunk 4 above)
Comment (kherock) [reply dist/lib/lib.js]: > TS won't understand it unless we use `require("cross-fetch").fetch` or `require("cross-fetch").default`

I don't think this is required, cross-fetch still exposes a CJS export regardless, and `require` semantics don't change regardless of the TypeScript version being used. However, I didn't realize cross-fetch exposes `fetch` as a named export. I'll update it since I prefer to use those anyway.

> what is the reason for switching to cross-fetch? I see your point for not polluting the global namespace, but in old browsers without fetch support that is exactly what we should do.

Since this library uses webpack, there shouldn't ever be much reason to put more than is actually necessary into the global namespace. cross-fetch does leverage the native implementations if they're available, but it won't put them onto `window`. If a project wants to use `fetch` on older browsers outside of the API provided by this library, I think it's more appropriate that they bring their in own polyfill and use the "pure" build instead.

Comment (kherock) [reply dist/lib/lib.js]: I don't think my changes impact that behavior anyway. With Webpack I don't think evaluating `require('whatwg-fetch')` is capable of mutating `window` anyway.

File: dist/lib/Client.js (hunk 3 above)
Comment (vlad-ignatov) [reply dist/lib/Client.js]: I was talking about the TS validation service that runs in my VS Code editor, even on JS files. Anyway, I found that simple `// @ts-ignore` comment is good enough to fix that warning.

**Comment by vlad-ignatov** [comment]
What you did here is a big change! I definitely want to merge some of it, but it is difficult to do that in one big PR. That is why I've decided to re-implement your changes one by one in new branch and test them carefully. This helped me discover some issues which I will post as separate comments, just FYI and to see if you have some ideas...

Switching to ESM is very important for us because at some point we want to switch to TypeScript and this will make that easier. Thank you for all you did to make that happen!

As for the polyfill-ing behavior of this library, I completely support your position, and yet I will have to disagree for backwards compatibility reasons. It can be used in Node or as a module, but the primary use-case remains the EHR. It must work in IE 10 and 11 so that people can deploy their apps to the EHR easily. The old versions of this library used to work out of the box thanks to jQuery (which was included in the build). We cannot tell people that their apps might break, unless they include certain polyfills.  That is why we can have a pure build for modern browsers (or Node build), but we have to keep the polyfills for IE. Note that our current goal is to make it easy for the existing SMART apps to update to the latest version of the client library. 

I will let you know when new branch (esm) is ready for merging, in case you want to check it out.

File: webpack.config.js [hunk 5]
```diff
@@ -39,36 +27,25 @@ module.exports = function(env, argv) {
...
-                                            proposals: true
-                                        },
-                                        // debug: true,
-                                        loose: true, // needed for IE 10
-                                    }
-                                ]
-                            ]
-                        }
-                    }
+                    use:  "babel-loader"
```
Comment (vlad-ignatov) [inline webpack.config.js]: This does not work well for me. It looks like the configuration from `.babelrc.js` is not picked up by `babel-loader` so had to use `use: "babel-loader?configFile=./.babelrc.js"` instead.

File: .babelrc.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+                        "last 2 firefox versions",
+                        "last 2 Edge versions",
+                        "ie 10-11"
+                    ],
+                    // debug: true,
+                    loose: true, // needed for IE 10
+                }]
+            ]
+        },
+        development: {
```
Comment (vlad-ignatov) [inline .babelrc.js]: Why did you remove the `@babel/plugin-transform-runtime` plugin. I used it to make the bundle smaller but maybe you had a good reason to remove it?

File: webpack.config.js [hunk 7]
```diff
@@ -39,36 +27,25 @@ module.exports = function(env, argv) {
                 {
                     test: /\.js$/,
                     include: [
-                        path.resolve(__dirname, "src"),
-                        path.resolve(__dirname, "node_modules/debug")
+                        path.join(__dirname, "src"),
+                        require.resolve("debug")
```
Comment (vlad-ignatov) [inline webpack.config.js]: This does not work properly. We need to tell Babel to transpile the entire `debug` source and not just the entry file.
Otherwise [some ES6 code](https://github.com/kherock/client-js/blob/master/dist/build/fhir-client.js#L7778-L7784) will leak into to bundle and break in IE.

File: .babelrc.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+            presets: [
+                ["@babel/preset-env", {
+                    useBuiltIns: "usage",
+                    modules: "commonjs",
+                    corejs: {
+                        version: 3
+                    },
+                    targets: "maintained node versions"
```
Comment (vlad-ignatov) [inline .babelrc.js]: This will polyfill many things (even promises) but we specify in `package.json` and in docs that Node should be >= 10. I know that Node 10 does support everything that we use and we shouldn't even need `useBuiltIns`, but just in case, I would use `targets: "node 10"` here.

File: .babelrc.js (hunk 6 above)
Comment (kherock) [reply .babelrc.js]: I didn't see any size reduction when I was inspecting the generated code with and without the plugin. The bundle is actually ~5K bigger with the transform.

File: webpack.config.js (hunk 7 above)
Comment (kherock) [reply webpack.config.js]: That makes sense. When I saw `node_modules/` I assumed it was intending to match relative to the project workspace.

File: webpack.config.js (hunk 5 above)
Comment (kherock) [reply webpack.config.js]: It definitely works for me. Maybe try using `babel.config.js` instead? Otherwise I'd prefer to see the object syntax, like
```
use: {
    loader: "babel-loader",
    options: {
        configFile: path.join(__dirname, ".babelrc.js")
    }
}
```

File: webpack.config.js (hunk 7 above)
Comment (kherock) [reply webpack.config.js]: Actually looking at this again, my code is equivalent to the previous revision!
```
Welcome to Node.js v12.13.0.
Type ".help" for more information.
> __dirname = process.cwd()
'/home/smart-on-fhir/client-js'
> path.resolve(__dirname, "node_modules/debug")
'/home/smart-on-fhir/client-js/node_modules/debug'
> path.resolve(__dirname, "node_modules/debug")
'/home/smart-on-fhir/client-js/node_modules/debug'
> 
```

If you want the behavior of including all versions in the hierarchy, `__dirname` shouldn't be there.

Comment (kherock) [reply webpack.config.js]: Also, core-js doesn't provide a console polyfill at all, so the line in question doesn't actually do anything. If you want to avoid surprises with console on IE in standards mode, you need your own polyfill, or something like [ie-console-patch](https://www.npmjs.com/package/ie-console-patch).

**Comment by kherock** [comment]
> I definitely want to merge some of it, but it is difficult to do that in one big PR.

I could have split this up into multiple, but I see you've already begun adding some of them to the main branch. I've rebased my branch on top of those, so at this point the diff should only contain changes that were reverted in your branch.

**Comment by vlad-ignatov** [comment]
In fact, I have already released `v2.1.1` on NPM. It includes most of your changes and I hope you can give it a try when you have time and verify that it works as expected.

It includes even newer versions of most dependencies (and npm audit fixes), so the build behavior might be a bit different. However, I think it is probably not worth spending too much time on build details, not that we are switching to TypeScript and all that is going to change anyway.

Thanks again for this significant contribution!

**Comment by vlad-ignatov** [comment]
This manually merged long time ago. I forgot to close the PR back then, so I am doing it now. Thanks a lot! It was instrumental for `v2`.
